File: Pemograman Berorientasi Objek/Praktikum 03/Soal 2/Matriks.cpp
#include "Matriks.h"
#include <iostream>

using namespace std;

Matriks::Matriks():num_brs(DEFSIZE),num_kol(DEFSIZE){
  b = new int*[num_brs];
  for (int i=0;i<num_brs;i++){
    b[i] = new int[num_kol];
    for (int j=0;j<num_kol;j++){
      b[i][j]=0;
    }
  }
}

Matriks::Matriks(int n, int m):num_brs(n),num_kol(m){
  b = new int*[num_brs];
  for (int i=0;i<num_brs;i++){
    b[i] = new int[num_kol];
    for (int j=0;j<num_kol;j++){
      b[i][j]=0;
    }
  }
}

Matriks::Matriks(const Matriks &m):num_brs(m.num_brs),num_kol(m.num_kol){
  b = new int*[num_brs];
  for (int i=0;i<num_brs;i++){
    b[i] = new int[num_kol];
    for (int j=0;j<num_kol;j++){
      b[i][j]=m.b[i][j];
    }
  }
}

Matriks::~Matriks(){
  for (int i=0;i<num_brs;i++){
    delete b[i];
  }
  delete b;
}

bool IsEqSize(Matriks m1, Matriks m2){
  return m1.num_brs== m2.num_brs && m1.num_kol== m2.num_kol;
}

Matriks Matriks::operator+(Matriks m2){
  int n,m;
  if (num_brs > m2.num_brs){
    n=num_brs;
  }
  else{
    n=m2.num_brs;
  };
  if (num_kol > m2.num_kol){
    m=num_kol;
  }
  else{
    m=m2.num_kol;
  };
  Matriks mx(n,m);
  for (int i=0;i<n;i++){
    for (int j=0;j<m;j++){
      if (i<num_brs && j<num_kol){
        mx.b[i][j]+=b[i][j];
      }
      if (i<m2.num_brs && j<m2.num_kol){
        mx.b[i][j]+=m2.b[i][j];
      }
    }
  }
  return mx;
}

Matriks operator-(const Matriks &m1, const Matriks &m2){
  int n,m;
  if (m1.num_brs > m2.num_brs){
    n=m1.num_brs;
  }
  else{
    n=m2.num_brs;
  };
  if (m1.num_kol > m2.num_kol){
    m=m1.num_kol;
  }
  else{
    m=m2.num_kol;
  };
  Matriks mx(n,m);
  for (int i=0;i<n;i++){
    for (int j=0;j<m;j++){
      if (i<m1.num_brs && j<m1.num_kol){
        mx.b[i][j]+=m1.b[i][j];
      }
      if (i<m2.num_brs && j<m2.num_kol){
        if (!(i<m1.num_brs && j<m1.num_kol)){
          mx.b[i][j]+=m2.b[i][j];
        }
        else{
          mx.b[i][j]-=m2.b[i][j];
        }
      }
    }
  }
  return mx;
}

Matriks Matriks::operator*(const int n){
  Matriks mx(num_brs,num_kol);
  for (int i=0;i<num_brs;i++){
    for (int j=0;j<num_kol;j++){
      mx.b[i][j]=b[i][j]*n;
    }
  }
  return mx;
}

Matriks operator*(const int n, const Matriks m2){
  Matriks mx(m2.num_brs,m2.num_kol);
  for (int i=0;i<m2.num_brs;i++){
    for (int j=0;j<m2.num_kol;j++){
      mx.b[i][j]=m2.b[i][j]*n;
    }
  }
  return mx;
}

void Matriks::SetData(int i, int j, int v){
  b[i][j]=v;
}

int Matriks::GetData(int i, int j) const{
  return b[i][j];
}

int Matriks::GetBrs() const{
  return num_brs;
}

int Matriks::GetKol() const{
  return num_kol;
}

bool Matriks::IsSimetri() const{
  return num_brs==num_kol;
}

File: Pemograman Berorientasi Objek/Praktikum 08/Soal 1/Matriks.java
/* NIM/Nama	 : 13517137/Vincent Budianto
 * Nama File : Matriks.java
 * Topik     : Praktikum 08
 * Tanggal   : 21 Maret 2019
 * Deskripsi : Class Matriks */

import java.io.*;
import java.util.*;
import java.lang.*;

class Matriks
{
	private int baris;
	private int kolom;
	private int data[][];
	
	public Matriks()
	{
		this.baris = 0;
		this.kolom = 0;
		this.data = new int[baris][kolom];
	}
	
	public Matriks(int baris, int kolom)
	{
		int i, j;
		
		this.baris = baris;
		this.kolom = kolom;
		this.data = new int[baris][kolom];
		
		for(i = 0; i < this.baris; i++)
		{
			for(j = 0; j < this.kolom; j++)
			{
				this.data[i][j] = 0;
			}
		}
	}
	
	public int getBrs()
	{
		return this.baris;
	}
	
	public int getKol()
	{
		return this.kolom;
	}
	
	public int getData(int baris, int kolom)
	{
		return (this.data[baris][kolom]);
	}
	
	public void setData(int baris, int kolom, int data)
	{
		this.data[baris][kolom] = data;
	}
	
	public boolean isEqSize(Matriks M)
	{
		return ((this.baris == M.baris) && (this.kolom == M.kolom));
	}
	
	public Matriks tambah(Matriks M)
	{
		int i, j, x, y;
		
		if (this.baris > M.baris)
		{
			x = this.baris;
		}
		else
		{
			x = M.baris;
		}
		
		if (this.kolom > M.kolom)
		{
			y = this.kolom;
		}
		else
		{
			y = M.kolom;
		}
		
		Matriks newM = new Matriks(x, y);

		for (i = 0; i < x; i++)
		{
			for (j = 0; j < y; j++)
			{
				if ((i < this.baris) && (j < this.kolom))
				{
					newM.data[i][j] += this.data[i][j];
				}
				
				if ((i < M.baris) && (j < M.kolom))
				{
					newM.data[i][j] += M.data[i][j];
				}
			}
		}
		
		return newM;
	}

	public Matriks kurang(Matriks M)
	{
		int i, j, x, y;
		
		if (this.baris > M.baris)
		{
			x = this.baris;
		}
		else
		{
			x = M.baris;
		}
		
		if (this.kolom > M.kolom)
		{
			y = this.kolom;
		}
		else
		{
			y = M.kolom;
		}
		
		Matriks newM = new Matriks(x, y);

		for (i = 0; i < x; i++)
		{
			for (j = 0; j < y; j++)
			{
				if ((i < this.baris) && (j < this.kolom))
				{
					newM.data[i][j] += this.data[i][j];
				}
				
				if ((i < M.baris) && (j < M.kolom))
				{
					if ((i < this.baris) && (j < this.kolom))
					{
						newM.data[i][j] -= M.data[i][j];
					}
					else
					{
						newM.data[i][j] += M.data[i][j];
					}
				}
			}
		}
		
		return newM;
	}
	
	public Matriks kali(int a)
	{
		int i, j;
		
		Matriks newM = new Matriks(this.baris, this.kolom);
		
		for (i = 0; i < newM.baris; i++)
		{
			for (j = 0; j < newM.kolom; j++)
			{
				newM.data[i][j] = this.data[i][j] * a;
			}
		}
		
		return newM;
	}
	
	public boolean isSimetri()
	{
		return (this.baris == this.kolom);
	}
}
